liblastlog2/src/lastlog2.c

```diff
@@ -0,0 +1,574 @@
+		if (error)
+			if (asprintf (error, "Failed to execute statement: %s",
+				      sqlite3_errmsg (db)) < 0)
+				*error = strdup ("Out of memory");
+
+		return -1;
+	}
+
+	if (sqlite3_bind_text (res, 1, user, -1, SQLITE_STATIC) != SQLITE_OK) {
+		if (error)
+			if (asprintf (error, "Failed to create replace statement for user: %s",
+				      sqlite3_errmsg (db)) < 0)
+				*error = strdup("Out of memory");
+
+		sqlite3_finalize(res);
+		return -1;
+	}
+
+	if (sqlite3_bind_int64 (res, 2, ll_time) != SQLITE_OK) {
+		if (error)
+			if (asprintf (error, "Failed to create replace statement for ll_time: %s",
+				      sqlite3_errmsg (db)) < 0)
+				*error = strdup("Out of memory");
+
+		sqlite3_finalize(res);
+		return -1;
+	}
+
+	if (sqlite3_bind_text (res, 3, tty, -1, SQLITE_STATIC) != SQLITE_OK) {
+		if (error)
+			if (asprintf (error, "Failed to create replace statement for tty: %s",
+				      sqlite3_errmsg (db)) < 0)
+				*error = strdup("Out of memory");
+
+		sqlite3_finalize(res);
+		return -1;
+	}
+
+	if (sqlite3_bind_text (res, 4, rhost, -1, SQLITE_STATIC) != SQLITE_OK) {
+		if (error)
+			if (asprintf (error, "Failed to create replace statement for rhost: %s",
+				      sqlite3_errmsg (db)) < 0)
+				*error = strdup("Out of memory");
+
+		sqlite3_finalize(res);
+		return -1;
+	}
+
+	if (sqlite3_bind_text (res, 5, pam_service, -1, SQLITE_STATIC) != SQLITE_OK) {
+		if (error)
+			if (asprintf (error, "Failed to create replace statement for PAM service: %s",
+				      sqlite3_errmsg (db)) < 0)
+				*error = strdup("Out of memory");
+
+		sqlite3_finalize(res);
+		return -1;
+	}
+
+	int step = sqlite3_step (res);
+
+	if (step != SQLITE_DONE) {
+		if (error)
+			if (asprintf (error, "Delete statement did not return SQLITE_DONE: %d",
+				      step) < 0)
+				*error = strdup("Out of memory");
+
+		sqlite3_finalize(res);
+		return -1;
+	}
+
+	sqlite3_finalize(res);
+
+	return 0;
+}
+
+/* Write a new entry. Returns 0 on success, -1 on failure. */
+int
+ll2_write_entry (const char *lastlog2_path, const char *user,
+		 int64_t ll_time, const char *tty, const char *rhost,
+		 const char *pam_service, char **error)
+{
+	sqlite3 *db;
+	int retval;
+
+	if ((db = open_database_rw (lastlog2_path, error)) == NULL)
+		return -1;
+
+	retval = write_entry (db, user, ll_time, tty, rhost, pam_service, error);
+
+	sqlite3_close (db);
+
+	return retval;
+}
+
+/* Write a new entry with updated login time.
+   Returns 0 on success, -1 on failure. */
+int
+ll2_update_login_time (const char *lastlog2_path, const char *user,
+		       int64_t ll_time, char **error)
+{
+	sqlite3 *db;
+	int retval;
+	char *tty;
+	char *rhost;
+	char *pam_service;
+
+	if ((db = open_database_rw (lastlog2_path, error)) == NULL)
+		return -1;
+
+	if (read_entry (db, user, 0, &tty, &rhost, &pam_service, error) != 0) {
+		sqlite3_close (db);
+		return -1;
+	}
+
+	retval = write_entry (db, user, ll_time, tty, rhost, pam_service, error);
+
+	sqlite3_close (db);
+
+	if (tty)
```

Comment:
As mentioned before `free()` can handle `NULL` just fine everywhere.

Reply:
I have removed all these checks in this file.